From: Mark Adler <madler@alumni.caltech.edu>
Date: Sun, 2 Feb 2025 21:34:20 +0100
Subject: [PATCH] Fix bug in undefer_input() that misplaced the input state.

    Fix bug in undefer_input() that misplaced the input state.


Imported from unzip_6.0-28.debian.tar.xz

Signed-off-by: Michael Olbrich <m.olbrich@pengutronix.de>
---
 fileio.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/fileio.c b/fileio.c
index c10ff63c5697..6290824b23df 100644
--- a/fileio.c
+++ b/fileio.c
@@ -532,8 +532,10 @@ void undefer_input(__G)
          * This condition was checked when G.incnt_leftover was set > 0 in
          * defer_leftover_input(), and it is NOT allowed to touch G.csize
          * before calling undefer_input() when (G.incnt_leftover > 0)
-         * (single exception: see read_byte()'s  "G.csize <= 0" handling) !!
+         * (single exception: see readbyte()'s  "G.csize <= 0" handling) !!
          */
+        if (G.csize < 0L)
+            G.csize = 0L;
         G.incnt = G.incnt_leftover + (int)G.csize;
         G.inptr = G.inptr_leftover - (int)G.csize;
         G.incnt_leftover = 0;
